esphome/components/remote_base/daikin_protocol.cpp

```diff
@@ -0,0 +1,110 @@
+#include "daikin_protocol.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace remote_base {
+
+static const char *const TAG = "remote.daikin";
+
+// IR Transmission
+static const uint32_t IR_FREQUENCY = 38000;
+static const uint32_t HEADER_MARK = 3360;
+static const uint32_t HEADER_SPACE = 1760;
+static const uint32_t BIT_MARK = 520;
+static const uint32_t ONE_SPACE = 1370;
+static const uint32_t ZERO_SPACE = 360;
+static const uint32_t MESSAGE_SPACE = 32300;
+
+static const uint8_t HEADER[] = {0x11, 0xda, 0x27, 0x00};
+
+void DaikinProtocol::encode(RemoteTransmitData *dst, const DaikinData &data) {
+  dst->set_carrier_frequency(IR_FREQUENCY);
+
+  dst->mark(HEADER_MARK);
+  dst->space(HEADER_SPACE);
+
+  uint8_t checksum = 0;
+
+  for (uint8_t byte : HEADER) {
+    DaikinProtocol::encode_byte(dst, byte);
+    checksum += byte;
+  }
+
+  for (uint8_t byte : data.data) {
+    DaikinProtocol::encode_byte(dst, byte);
+    checksum += byte;
+  }
+
+  DaikinProtocol::encode_byte(dst, checksum);
+
+  dst->mark(BIT_MARK);
+  dst->space(MESSAGE_SPACE);
+}
+
+optional<DaikinData> DaikinProtocol::decode(RemoteReceiveData src) {
+  if (!src.expect_item(HEADER_MARK, HEADER_SPACE)) {
+    return {};
+  }
+
+  DaikinData out;
+  bool end_message = false;
+
+  while (src.is_valid() && !end_message) {
+    // Fetch each byte
+    uint8_t byte = 0;
+    for (int8_t bit = 0; bit < 8; bit++) {
+      if (src.expect_item(BIT_MARK, ONE_SPACE)) {
+        byte |= 1 << bit;
+      } else if (!src.expect_item(BIT_MARK, ZERO_SPACE)) {
+        end_message = true;
+        break;
+      }
+    }
+
+    if (!end_message)
+      out.data.push_back(byte);
+  }
+
+  if (out.data.size() <= 4)
+    return {};
+
+  // Calculate checksum and check header
+  uint8_t checksum = 0, i = 0;
+  for (auto it = out.data.begin(); it != std::prev(out.data.end()); i++) {
+    checksum += *it;
+
+    if (i < sizeof(HEADER)) {
+      if (*it != HEADER[i]) {
+        ESP_LOGD(TAG, "Invalid header byte (%hhu): received %02X, expected %02X", i, *it, HEADER[i]);
```

Comment:
```suggestion
        ESP_LOGD(TAG, "Invalid header byte (%" PRIu8 "): received %02X, expected %02X", i, *it, HEADER[i]);
```

```suggestion
        ESP_LOGD(TAG, "Invalid header byte (%u): received %02X, expected %02X", i, *it, HEADER[i]);
```

Size prefix hh is used only for scanf Afaik, in printf, before printing everything gets cast to largest integer. (it's best to check docs for each platform or just use format macro specifier to make it compatible across multiple platforms)